File: windows/MicroKeys.Core/FrameDecoder.cs
using System.Text;

namespace MicroKeys.Core;

/// <summary>
/// Reassembles JSON-RPC messages from a stream of vendor HID input reports.
/// Each report is <c>[0x02][len][utf-8 json…]</c>; Windows prefixes the buffer
/// with the report id (0x06), other paths may not, and both layouts are accepted.
/// Messages are CRLF-terminated, may span several reports, and several may
/// share one report.
/// </summary>
public sealed class FrameDecoder
{
    public const byte ReportId = 6;
    public const byte OpcodeData = 0x02;

    private readonly List<byte> _buffer = new();

    /// <summary>Consume one raw report; return every event it completed.</summary>
    public List<PadEvent> Feed(ReadOnlySpan<byte> raw)
    {
        var events = new List<PadEvent>();
        int start;
        if (raw.Length >= 3 && raw[0] == ReportId && raw[1] == OpcodeData) start = 1;
        else if (raw.Length >= 2 && raw[0] == OpcodeData) start = 0;
        else return events;

        int length = raw[start + 1];
        int bodyStart = start + 2;
        int bodyEnd = Math.Min(raw.Length, bodyStart + length);
        if (bodyEnd <= bodyStart) return events;
        _buffer.AddRange(raw[bodyStart..bodyEnd].ToArray());

        int crlf;
        while ((crlf = IndexOfCrlf()) >= 0)
        {
            var line = _buffer.GetRange(0, crlf).ToArray();
            _buffer.RemoveRange(0, crlf + 2);
            string text;
            try { text = Encoding.UTF8.GetString(line).Trim(); }
            catch (ArgumentException) { continue; }
            if (text.Length == 0) continue;
            var e = PadEvent.Parse(text);
            if (e is not null) events.Add(e);
        }
        // A runaway buffer means we lost framing; drop it rather than grow forever.
        if (_buffer.Count > 8192) _buffer.Clear();
        return events;
    }

    public void Reset() => _buffer.Clear();

    private int IndexOfCrlf()
    {
        for (int i = 0; i + 1 < _buffer.Count; i++)
            if (_buffer[i] == 0x0D && _buffer[i + 1] == 0x0A) return i;
        return -1;
    }
}

File: macos/Sources/MicroKeysCore/FrameDecoder.swift
import Foundation

/// Reassembles JSON-RPC messages from a stream of vendor HID input reports.
///
/// Wire facts (see docs/PROTOCOL.md in the freemicro project):
/// * Each report is `[0x02][len][utf-8 json…]`. macOS may or may not prefix the
///   buffer with the report id (`0x06`); we accept both layouts.
/// * Messages are CRLF-terminated. One message can span several reports and one
///   report can carry the tail of one message plus the start of the next.
public struct FrameDecoder {
    public static let reportID: UInt8 = 6
    public static let opcodeData: UInt8 = 0x02

    private var buffer: [UInt8] = []

    public init() {}

    /// Consume one raw report; return every event it completed.
    public mutating func feed(_ raw: [UInt8]) -> [PadEvent] {
        let start: Int
        if raw.count >= 3, raw[0] == Self.reportID, raw[1] == Self.opcodeData {
            start = 1
        } else if raw.count >= 2, raw[0] == Self.opcodeData {
            start = 0
        } else {
            return []
        }
        let length = Int(raw[start + 1])
        let bodyStart = start + 2
        let bodyEnd = min(raw.count, bodyStart + length)
        guard bodyEnd > bodyStart else { return [] }
        buffer.append(contentsOf: raw[bodyStart..<bodyEnd])

        var events: [PadEvent] = []
        while let lineEnd = indexOfCRLF() {
            let line = Array(buffer[..<lineEnd])
            buffer.removeSubrange(..<(lineEnd + 2))
            guard let text = String(bytes: line, encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines),
                !text.isEmpty
            else { continue }
            if let event = PadEvent.parse(text) {
                events.append(event)
            }
        }
        // A runaway buffer means we lost framing; drop it rather than grow forever.
        if buffer.count > 8192 { buffer.removeAll() }
        return events
    }

    public mutating func reset() { buffer.removeAll() }

    private func indexOfCRLF() -> Int? {
        guard buffer.count >= 2 else { return nil }
        for i in 0..<(buffer.count - 1) where buffer[i] == 0x0D && buffer[i + 1] == 0x0A {
            return i
        }
        return nil
    }
}
